backend/pubsub.js

```diff
@@ -322,9 +336,50 @@ const publicMethods = {
   ) {
     const server = this
 
+    const msg = typeof message === 'string' ? message : JSON.stringify(message)
+    let shortMsg
+    // Utility function to remove `data` (i.e., the GIMessage data) from a
+    // message. We need this for push notifications, which may have a certain
+    // maximum size (usually around 4 KiB)
+    const shortenPayload = () => {
+      if (!shortMsg && (typeof message === 'object' && message.type === NOTIFICATION_TYPE.ENTRY && message.data)) {
+        delete message.data
+        shortMsg = JSON.stringify(message)
+      }
+      return shortMsg
+    }
+
     for (const client of to || server.clients) {
+      if (client.endpoint) {
+        // The max length for push notifications in many providers is 4 KiB.
+        // However, encrypting adds a slight overhead of 17 bytes at the end
+        // and 86 bytes at the start.
+        if (msg.length > (4096 - 86 - 17)) {
+          if (!shortenPayload()) {
+            console.info('Skipping too large of a payload for', client.id)
+            continue
+          }
+        }
+        postEvent(client, shortMsg || msg).catch(e => {
+          // If we have an error posting due to too large of a payload and the
+          // message wasn't already shortened, try again
+          if (e?.message === 'Payload too large') {
+            if (shortMsg || !shortenPayload()) {
+              // The max length for push notifications in many providers is 4 KiB.
+              console.info('Skipping too large of a payload for', client.id)
+              return
+            }
+            postEvent(client, shortMsg).catch(e => {
+              console.error(e, 'Error posting push notification')
+            })
+            return
+          }
+          console.error(e, 'Error posting push notification')
+        })
+        continue
+      }
       if (client.readyState === WebSocket.OPEN && client !== except) {
-        client.send(typeof message === 'string' ? message : JSON.stringify(message))
+        client.send(msg)
```

Comment:
Two questions:

1. where exactly do you say "turn off the push notifications endpoint because the user has opened the app and has the websocket open"?

2. Does it make sense to have the message be sent out to both the websocket and via push as is currently being done (if both are available)? Or should this `if` be moved above the other if above, and the other if be made an else-if?

And finally, could you add a comment at the top of the for-loop with your answer?

Reply:
1. This function doesn't "turn off the push notifications endpoint because the user has opened the app and has the websocket open", that's done when a WS connection is open and web push information is sent.
2. It doesn't make sense to send the message twice, and this function doesn't.

Reply:
> It doesn't make sense to send the message twice, and this function doesn't.

Can you explain in more detail why? It seems like both `if` statements can be true at the same time to me.

Reply:
Are you talking about:

> if (client.endpoint) {
> if (client.readyState === WebSocket.OPEN && client !== except) {

They can't be both be true "at the same time", although there could be a client for which there's both a WS and a web push subscription (which would be separate clients). As stated, the logic to prevent sending duplicate events in this situation is in the handler when receiving push subscription information and not here. This function doesn't concern itself with deduplication.